src/components/Canvas/Canvas.js

```diff
@@ -145,67 +35,208 @@ const Canvas = ({ infocanvas, infoimage, elements, ...props }) => {
+      y += infoimage.offsetY;
+    }
 
-    render();
+    return { sourceX, sourceY, sourceWidth, sourceHeight, x, y, width, height };
+  };
 
-    return () => {
-      window.cancelAnimationFrame(animationFrameId);
-    };
-  }, [draw]);
+  const centerImage = (pic, canvasWidth, canvasHeight) => {
+    const imageWidth = pic.width;
+    const imageHeight = pic.height;
+
+    const canvasAspectRatio = canvasWidth / canvasHeight;
+    const imageAspectRatio = imageWidth / imageHeight;
+
+    let sourceX, sourceY, sourceWidth, sourceHeight;
+
+    if (imageAspectRatio > canvasAspectRatio) {
+      sourceWidth = imageHeight * canvasAspectRatio;
+      sourceHeight = imageHeight;
+      sourceX = (imageWidth - sourceWidth) / 2;
+      sourceY = 0;
+    } else {
+      sourceWidth = imageWidth;
+      sourceHeight = imageWidth / canvasAspectRatio;
+      sourceX = 0;
+      sourceY = (imageHeight - sourceHeight) / 2;
+    }
+
+    return { sourceX, sourceY, sourceWidth, sourceHeight };
+  };
 
+  const drawWrappedText = (
+    ctx,
+    text,
+    x,
+    y,
+    blockWidth,
+    lineHeight,
+    fontHeight,
+    alignment
+  ) => {
+    const words = text.split(' ');
+    let line = '';
+    blockWidth = blockWidth ?? ctx.canvas.width;
+
+    for (let i = 0; i < words.length; i++) {
+      let testLine = line + words[i];
+      const metrics = ctx.measureText(testLine);
+      const testWidth = metrics.width;
+      testLine += ' ';
+
+      if (testWidth > blockWidth) {
+        let offsetX = 0;
+
+        switch (alignment) {
+          case 'center':
+            offsetX = (blockWidth - ctx.measureText(line).width) / 2;
+            break;
+          case 'right':
+            offsetX = blockWidth - ctx.measureText(line).width;
+            break;
+          default:
+            break;
+        }
+
+        ctx.fillText(line, x + offsetX, y + fontHeight);
+        line = words[i] + ' ';
+        y += lineHeight;
+      } else {
+        line = testLine;
+      }
+    }
+
+    let offsetX = 0;
+
+    switch (alignment) {
+      case 'center':
+        offsetX = (blockWidth - ctx.measureText(line).width) / 2;
+        break;
+      case 'right':
+        offsetX = blockWidth - ctx.measureText(line).width;
+        break;
+      default:
+        break;
+    }
+    ctx.fillText(line, x + offsetX, y + lineHeight);
+  };
+  const draw = async () => {
+    console.log('Drawing...');
+    const ctx = contextRef?.current;
+    if (!ctx) {
+      return;
+    }
+
+    if (infoimage.srcimage) {
+      await drawImageOnCanvas(
+        ctx,
+        infoimage.srcimage,
+        infoimage,
+        calculateImageParameters
+      );
+    }
+    drawElementsOnCanvas(
+      ctx,
+      elements.filter((style) => style.type === 'image'),
+      drawWrappedText
+    );
+
+    drawElementsOnCanvas(
+      ctx,
+      elements.filter((style) => style.type === 'text'),
+      drawWrappedText
+    );
+  };
+
+  useEffect(() => {
+    draw();
```

Comment:
Props, state, and other values declared inside the component are reactive because they’re calculated during rendering and participate in the React data flow.

All variables declared in the component body are reactive
Props and state aren’t the only reactive values. Values that you calculate from them are also reactive. If the props or state change, your component will re-render, and the values calculated from them will also change. This is why all variables from the component body used by the Effect should be in the Effect dependency list.

https://react.dev/learn/lifecycle-of-reactive-effects#effects-react-to-reactive-values попробуй еще раз вот с этим хорошо разобраться